core/src/main/java/io/questdb/cairo/wal/WalPurgeJob.java

```diff
@@ -0,0 +1,403 @@
+        return path.of(engine.getConfiguration().getRoot())
+                .concat(tableName).$();
+    }
+
+    private Path setSeqTxnPath(CharSequence tableName) {
+        return path.of(engine.getConfiguration().getRoot())
+                .concat(tableName)
+                .concat(Sequencer.SEQ_DIR).$();
+    }
+
+    private Path setTxnPath(CharSequence tableName) {
+        return path.of(engine.getConfiguration().getRoot())
+                .concat(tableName)
+                .concat(TableUtils.TXN_FILE_NAME).$();
+    }
+
+    private Path setWalPath(CharSequence tableName, int walId) {
+        return path.of(engine.getConfiguration().getRoot())
+                .concat(tableName).concat("wal").put(walId).$();
+    }
+
+    private Path setWalLockPath(CharSequence tableName, int walId) {
+        path.of(engine.getConfiguration().getRoot())
+                .concat(tableName).concat("wal").put(walId);
+        TableUtils.lockName(path);
+        return path;
+    }
+
+    private Path setSegmentPath(CharSequence tableName, int walId, int segmentId) {
+        return path.of(engine.getConfiguration().getRoot())
+                .concat(tableName).concat("wal").put(walId).slash().put(segmentId).$();
+    }
+
+    private Path setSegmentLockPath(CharSequence tableName, int walId, int segmentId) {
+        path.of(engine.getConfiguration().getRoot())
+                .concat(tableName).concat("wal").put(walId).slash().put(segmentId);
+        TableUtils.lockName(path);
+        return path;
+    }
+
+    private boolean couldObtainLock(Path path) {
+        final long lockFd = TableUtils.lock(ff, path, false);
+        if (lockFd != -1L) {
+            ff.close(lockFd);
+            return true;  // Could lock/unlock.
+        }
+        return false;  // Could not obtain lock.
+    }
+
+    private boolean walIsInUse(CharSequence tableName, int walId) {
+        return !couldObtainLock(setWalLockPath(tableName, walId));
+    }
+
+    private void discoverWalDirectoriesIter(long pUtf8NameZ, int type) {
+        if ((type == Files.DT_DIR) && matchesWalNamePattern(walName.of(pUtf8NameZ))) {
+            // We just record the name for now in a set which we'll remove items to know when we're done.
+            int walId = 0;
+            try {
+                walId = Numbers.parseInt(walName, 3, walName.length());
+            } catch (NumericException _ne) {
+                return;  // Ignore non-wal directory
+            }
+            discoveredWalIds.add(walId);
+            if (walIsInUse(tableName, walId)) {
+                walsInUse.add(walId);
+            }
+        }
+    }
+
+    private void discoverWalDirectories() {
+        ff.iterateDir(setTablePath(tableName), this::discoverWalDirectoriesIter);
+    }
+
+    private void populateWalInfoDataFrame() {
+        setTxnPath(tableName);
+        txReader.ofRO(path, PartitionBy.NONE);
+        final long lastAppliedTxn = txReader.unsafeReadVersion();
```

Comment:
Close `txReader` once you read all the info from it. It is safe to close and re-use it later